File: Ruby/segment_tree.rb
# returns the number of bits required to store n.
# n should be an integer.
def get_bit_length(n)
    if n == 0
        # we require no bit to store 0.
        # That is, absence of bit is considered 0.
        return 0
    else 
        return 1 + get_bit_length(n >> 1)
    end
end


def nearest_pow_2(n)
    # returns next number which is a power of 2.
    if n & (n - 1) == 0
        # n is already a power of 2.
        return n
    end

    # current number is not already power of 2.
    # answer is pow(2, ⌈log2(n)⌉)
    # number of bits required to represent n is ⌈log2(n)⌉.
    return 1 << get_bit_length(n)
end


class SegmentTree
    
    attr_accessor :internal_arr, :n_leaves, :func
    # full binary tree represented in array form.
    @internal_arr = nil

    @n_leaves = 0

    # bivariate function which determines 
    #    the type of query segment tree will handle.
    @func = nil

    # initializes segment tree with given array of objects.
    # arr: arr is a collection of objects. 
    #      arr should be an iterable which supports each function.
    # func: a bivariate function.
    #       must work with all objects in the collection provided (arr)
    # fallback: default value of the leaves which won't affect the 
    #           computation of bivariate func 
    #           when value of only one parameter is given.
    #    For example:
    #      In case if we are building tree to answer min queries,
    #          fallback = inf will always yield the right answer.
    #      If segment tree is required to answer max queries,
    #          fallback = -inf
    #      if we are answering sum queries,
    #          fallback = 0
    def initialize(arr, func, fallback)
        # check if the input array is an iterable 
        #       that supports each function.
        unless arr.respond_to? :each
            # raise TypeError, "The object must be an iterable."
        end
        
        @func = func
        @fallback = fallback

        self.build_tree(arr)
    end

    def build_tree(arr)
        # sets all the internal nodes of the segment tree.
        # This method requires func argument to be set.

        unless arr != nil
            raise ArgumentError, "Uninitialized array object (arr)"
        end

        # internal array will be a representation of a binary tree.
        # leaves of the tree is the array elements.
        # since segment tree is a full binary tree, 
        #     the number of leaves should be a power of 2.
        # Hence, for a lossless tree conversion, we take number of elements 
        #     as pow of 2 just >= the current arr size.
        #
        # If number of leaves in a full binary tree is n,
        #   total number of internal nodes in the tree is n - 1
        # Hence, total number of nodes in the tree will be 2n - 1
        # but we will use 2n elements and use a 1 based indexing.
        @n_leaves = nearest_pow_2(arr.size)
        n = @n_leaves << 1

        @internal_arr = Array.new(n, @fallback)
        # copying array to leaves of the tree.
        # ignore magic numbers 1 and 3 below. Will explain later.
        for i in (@n_leaves .. @n_leaves + arr.size - 1)
            @internal_arr[i] = arr[i - @n_leaves]
        end
 
        # updating all the nodes of the tree.
        (1 - @n_leaves .. -1).each do |i|
            i = -i
            @internal_arr[i] = @func.call(@internal_arr[i << 1], @internal_arr[i << 1 | 1])
        end
    end

    # updating the array element with new_val
    # idx in (0..arr.size)
    def update(idx, new_val)
        # idx is 1-based
        idx += @n_leaves - 1

        # Note: this will allow updating indices 
        #     that were not in original array  but is 
        #     less than next power of 2 of arr.size.
        unless idx < @internal_arr.size
            raise KeyError, "Index Out of Bounds Error"
        end
        @internal_arr[idx] = new_val
        
        p = idx >> 1
        
        # since we are using a 1-based indexing, 
        #     p=0 is an invalid state.
        while p > 0 do
            @internal_arr[p] = @func.call(@internal_arr[p << 1], @internal_arr[p << 1 | 1])
            p >>= 1
        end
    end

    # The most important part of segment tree.
    # returns @func(arr[l..r]) equivalent in O(log(n)) time.
    # Args:
    #    
    def query(
        query_l, query_r,  # left and right bounds of the query.
                           #     right inclusive.
        segment_l = 1,
        segment_r = @n_leaves, 
        i = 1
    )
        if query_l <= segment_l and segment_r <= query_r
            # current segment is subsumed by the query segment.
            return @internal_arr[i]
        elsif segment_r < query_l or query_r < segment_l
            # disjoint segments.
            return @fallback
        else
            mid = (segment_l + segment_r) >> 1
            return @func.call(
                query(
                    query_l, query_r,
                    segment_l, mid,
                    i << 1
                ),
                query(
                    query_l, query_r,
                    mid + 1, segment_r,
                    (i << 1) | 1
                )
            )
        end
    end
end




class ValueError < StandardError
end


def main
    # Function for demonstrating and testing the Segment Tree.

    arr = [1, 2, 3, 5, 5]
    add_func = Proc.new {|x, y| x + y}
    min_func = Proc.new {|x, y| [x, y][x < y ? 0 : 1]}
    
    
    # segment tree for answering range sum query:
    sum_st = SegmentTree.new(arr, func=add_func, fallback=0)
    raise ValueError, "Segment Tree Faulty" unless sum_st.query(1, 1) == 1
    raise ValueError, "Segment Tree Faulty" unless sum_st.query(2, 4) == arr[1..3].inject(0, :+)
    raise ValueError, "Segment Tree Faulty" unless sum_st.query(5, 4) == fallback
    sum_st.update(4, 4)
    raise ValueError, "Segment Tree Faulty" unless sum_st.query(1, 5) == 15   # 1 + 2 + 3 + 4 + 5 = 5 * 6 >> 1 = 15


    # segment tree for answering range minimum query:
    arr = [1, 2, 3, 5, 5]
    fallback = 1.0 / 0     # inf since min(obj, inf) = obj
    min_st = SegmentTree.new(arr, func=min_func, fallback)
    raise ValueError, "Segment Tree Faulty" unless min_st.query(1, 1) == 1
    raise ValueError, "Segment Tree Faulty" unless min_st.query(2, 5) == 2
    min_st.update(3, -1) # setting third element to -1
    raise ValueError, "Segment Tree Faulty" unless min_st.query(1, 5)== -1
end

File: Java/SegmentTree.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/** A class to represent a Segment Tree.
 */
public class SegmentTree {
    /**
     * @param {List} arr   The segment tree will be built using this iterable
     * @param {function} func  The function that will be used to built the tree
     * @param {number} def     The default value with which the tree array is initialized
     */
    List<Integer> arr;
    Operations func;
    int def, n_leaves, len_arr;
    List<Integer> tree;

    public SegmentTree(List<Integer> arr, Operations func, int def) {
        if (arr == null)
            throw new NullPointerException("Argument arr should not be null");
        this.arr = arr;
        this.len_arr = this.arr.size();

        if (func == null)
            throw new NullPointerException("Argument func should not be null");
        this.func = func;
        this.def = def;

        buildTree();
    }

    /**
     * Function to return the next number which is a power of 2.
     *
     * @param n
     * @returns number
     */
    private int next_power_of_2(int n) {
        // Checks if n is already a power of 2
        if ((n & n-1) == 0)
            return n;
        int val = 1;

        // Loop will run until it give us the value which is greater than n.
        while (val <= n)
            // val will be left shifted by 1 in each iteration.
            val <<= 1;

        return val;
    }

    /**Function to build the segment tree using the function provided.
     */
    private void buildTree() {
        // If number of leaves in a full binary tree is n,
        // total number of internal nodes in the tree is n - 1.
        // Hence, total number of nodes in the tree will be 2n - 1.
        this.n_leaves = next_power_of_2(this.len_arr);
        int n = this.n_leaves << 1;

        // initializing all nodes of the tree to 0
        this.tree = new ArrayList<>();
        for (int i = 0; i < n; i++)
            this.tree.add(this.def);

        // copying array to leaves of the tree
        for (int i = 0; i < this.len_arr; i++)
            this.tree.set(this.len_arr+i, this.arr.get(i));

        // updating all the nodes of the tree
        for (int i = this.len_arr - 1; i > 0; i--)
            this.tree.set(i, this.func.operate(this.tree.get(i << 1), this.tree.get(i << 1 | 1)));
    }

    /**
     * Function to update the element at the given index with new value
     *
     * @param i
     * @param value
     */
    public void update(int i, int value) {
        if (i > this.len_arr)
            throw new IndexOutOfBoundsException("Index i should be less than length of the arr");

        i += this.len_arr;
        this.tree.set(i, value);
        int p = i >> 1;

        while (p > 0) {
            this.tree.set(p, this.func.operate(this.tree.get(p << 1), this.tree.get(p << 1 | 1)));
            p >>= 1;
        }
    }

    /**
     * Returns the self.func(arr[l..r]) equivalent in O(log(n)) time.
     *
     * @param left
     * @param right
     * @returns number
     */
    public int query(int left, int right) {
        return query(left, right, 1, this.n_leaves, 1);
    }

    private int query(int left, int right, int l, int r, int i) {
        if (left < 0 || right > this.len_arr || left > right)
            throw new IndexOutOfBoundsException();

        if (r <= right && left <= l)
            return this.tree.get(i - 1);
        else if (r < left || right < l)
            return this.def;
        else {
            int mid = (l + r) >> 1;
            return this.func.operate(
                query(
                    left, right,
                    l, mid,
                    i << 1
                ),
                query(
                    left, right,
                    mid + 1, r,
                    i << 1 | 1
                )
            );
        }
    }

}

@FunctionalInterface
interface Operations {
    int operate(int x, int y);
}

// Sample test cases to test the implementation of Segment Tree
class App {
    public static void main(String[] args) {
        List<Integer> arr = new ArrayList<>(Arrays.asList(1, 2, 3, 5, 5));

        // Testing segment tree for sum of ranges
        Operations add = (x, y) -> x + y;
        SegmentTree sgt1 = new SegmentTree(arr, add, 0);
        assert sgt1.query(1, 3) == 10;
        sgt1.update(2, 5); // changing the element at index 2 to 5
        assert sgt1.query(1, 3) == 12;

        // Testing segment tree for multiplication of ranges
        Operations mult = (x, y) -> x * y;
        SegmentTree sgt2 = new SegmentTree(arr, mult, 1);
        assert sgt2.query(1, 3) == 30;
        sgt2.update(2, 5); // changing the element at index 2 to 5
        assert sgt2.query(1, 3) == 50;

    }
}
